### PR TITLE
gluon bce loss

### PR DESCRIPTION
**Pull request by szha**

### CODE REVIEW

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -142,6 +144,46 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
         return F.mean(loss, axis=self._batch_axis, exclude=True)
 
 
+class BinaryCrossEntropyLoss(Loss):
+    r"""The cross-entropy loss for binary classification.
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: why not just call it logistic regression loss. Is binaryce a standard name?

Comment (szha) [inline python/mxnet/gluon/loss.py]: [pytorch uses it](http://pytorch.org/docs/master/nn.html#bceloss)

File: python/mxnet/gluon/loss.py [hunk 2]
```diff
@@ -142,6 +144,46 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+        in the batch, `sample_weight` should have shape (64, 1).
+    batch_axis : int, default 0
+        The axis that represents mini-batch.
+    """
+    def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
+        super(BinaryCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
+        self._from_sigmoid = from_sigmoid
+
+    def hybrid_forward(self, F, output, label, sample_weight=None):
+        label = label.reshape((-1, 1))
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: see l1loss for how to reshape

File: python/mxnet/gluon/loss.py [hunk 3]
```diff
@@ -239,3 +281,59 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+        Global scalar weight for loss.
+    sample_weight : Symbol or None
+        Per sample weighting. Must be broadcastable to
+        the same shape as loss. For example, if loss has
+        shape (64, 10) and you want to weight each sample
+        in the batch, `sample_weight` should have shape (64, 1).
+    batch_axis : int, default 0
+        The axis in label that represents mini-batch.
+    """
+    def __init__(self, output_layout='NTC', weight=None, batch_axis=0, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: output_layout & label_layout

File: python/mxnet/gluon/loss.py [hunk 4]
```diff
@@ -239,3 +281,59 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+
+    See `"Connectionist Temporal Classification: Labelling Unsegmented
+    Sequence Data with Recurrent Neural Networks"
+    <http://www.cs.toronto.edu/~graves/icml_2006.pdf>`_ paper for more information.
+
+    The prediction output should be an activation vector without softmax, with shape
+    according to the output_layout:
+    **TNC**: *(sequence_length, batch_size, alphabet_size + 1)*
+    **NTC**: *(batch_size, sequence_length, alphabet_size + 1)*
+    **out**: *(batch_size)*.
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: separate section for out

File: python/mxnet/gluon/loss.py [hunk 5]
```diff
@@ -239,3 +281,59 @@ def hybrid_forward(self, F, output, label, sample_weight=None):
...
+    **batch_axis=0**: *(batch_size, label_sequence_length)*
+    **batch_axis=1**: *(label_sequence_length, batch_size)*
+
+    If a sequence of labels is shorter than *label_sequence_length*, use the special
+    padding character 0 at the end of the sequence to conform it to the correct
+    length. For example, if *label_sequence_length* = 4, and one has two sequences
+    of labels [2, 1] and [3, 2, 2], the resulting ```label``` tensor should be
+    padded to be::
+
+      [[2, 1, 0, 0], [3, 2, 2, 0]]
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: does it mean the labels are one based?

File: python/mxnet/gluon/loss.py (hunk 2 above)
Comment (szha) [inline python/mxnet/gluon/loss.py]: Got it. I see it happening three times in the code, so I will refactor this part to a utility function.

File: python/mxnet/gluon/loss.py (hunk 5 above)
Comment (szha) [inline python/mxnet/gluon/loss.py]: Yes, since 0 is special character. I will take a look at cudnn7 CTC too and see if we can make it consistent.

**Comment by szha** [comment]
Updating the interface now for potential integration with cudnn7. Do not merge yet.

File: python/mxnet/gluon/loss.py [hunk 6]
```diff
@@ -148,13 +150,48 @@ def __init__(self, weight=None, batch_axis=0, **kwargs):
...
+    """
+    def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
+        super(BinaryCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
+        self._from_sigmoid = from_sigmoid
+
+    def hybrid_forward(self, F, output, label, sample_weight=None):
+        label = _reshape_label_as_output(F, output, label)
+        if not self._from_sigmoid:
+            max_val = F.maximum(-output, 0)
+            loss = output - output*label + max_val + F.log(F.exp(-max_val)+F.exp(-output-max_val))
```
Comment (piiswrong) [inline python/mxnet/gluon/loss.py]: add test to compare against softmaxcrossentropy with 2 classes

Comment (szha) [inline python/mxnet/gluon/loss.py]: added
